orchagent/sfloworch.cpp

```diff
@@ -0,0 +1,674 @@
+
+    if (sample_obj ==  m_sflowRateSampleMap.end())
+    {
+        new_session.rate = rate;
+        if (!sflowCreateSession(new_session))
+        {
+            SWSS_LOG_ERROR("Creating sflow session with rate %d failed", rate);
+            return false;
+        }
+        m_sflowRateSampleMap[rate] = new_session.m_sample_id;
+        m_sflowSampleRefMap[new_session.m_sample_id] = 0;
+    }
+    else
+    {
+        new_session.m_sample_id = sample_obj->second;
+    }
+
+    if (old_session->second.adminState)
+    {
+        if (!sflowAddPort(new_session, port_id))
+        {
+            return false;
+        }
+    }
+
+    m_sflowSampleRefMap[new_session.m_sample_id]++;
+    m_sflowSampleRefMap[old_session->second.m_sample_id]--;
+
+    if (m_sflowSampleRefMap[old_session->second.m_sample_id] == 0)
+    {
+        if (!sflowDestroySession(old_session->second))
+        {
+            SWSS_LOG_ERROR("Failed to clean old session %lx",
+                           old_session->second.m_sample_id);
+        }
+        else
+        {
+            m_sflowRateSampleMap.erase(old_session->second.rate);
+            m_sflowSampleRefMap.erase(old_session->second.m_sample_id);
+        }
+    }
+    new_session.globalConfigured = old_session->second.globalConfigured;
+    new_session.adminState = old_session->second.adminState;
+
+    m_sflowPortSessionMap[port_id] = new_session;
+    return true;
+}
+
+bool SflowOrch::sflowAddPort(SflowSession &session, sai_object_id_t port_id)
+{
+    sai_attribute_t attr;
+    sai_status_t    sai_rc;
+
+    if (!sflowStatus)
+    {
+        return true;
+    }
+
+    attr.id = SAI_PORT_ATTR_INGRESS_SAMPLEPACKET_ENABLE;
+    attr.value.oid = session.m_sample_id;
+
+    sai_rc = sai_port_api->set_port_attribute(port_id, &attr);
+
+    if (sai_rc != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to set session %lx on port %lx",
+                       session.m_sample_id, port_id);
+        return false;
+    }
+    return true;
+}
+
+bool SflowOrch::sflowDelPort(SflowSession &session, sai_object_id_t port_id)
+{
+    sai_attribute_t attr;
+    sai_status_t    sai_rc;
+
+
+    if (!sflowStatus)
+    {
+        return true;
+    }
+
+    attr.id = SAI_PORT_ATTR_INGRESS_SAMPLEPACKET_ENABLE;
+    attr.value.oid = SAI_NULL_OBJECT_ID;
+
+    sai_rc = sai_port_api->set_port_attribute(port_id, &attr);
+
+    if(sai_rc != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to delete session %lx on port %lx",
+                       session.m_sample_id, port_id);
+        return false;
+    }
+    return true;
+}
+
+bool SflowOrch::sflowGlobalConfigure(bool enable)
+{
+    bool ret = true;
+
+    for(auto& pair: gPortsOrch->getAllPorts())
+    {
+        auto& port = pair.second;
+        if (port.m_type != Port::PHY) continue;
+
+        auto sflowInfo = m_sflowPortSessionMap.find(port.m_port_id);
+        if(sflowInfo == m_sflowPortSessionMap.end())
+        {
+            if(!enable)
```

Comment:
This is inside another if condition inside for loop. Only when there is no session available for port and when global is enabled, the session needs to be created. When enable is set to false when there is no session on port already it can be skipped.